build-system/tasks/clean.js

```diff
@@ -3,75 +3,80 @@
 const argv = require('minimist')(process.argv.slice(2));
 const del = require('del');
 const fs = require('fs-extra');
+const globby = require('globby');
 const path = require('path');
 const {cyan, yellow} = require('kleur/colors');
+const {ignoreFile, splitIgnoreListByHeader} = require('./check-ignore-list');
 const {log} = require('../common/logging');
 
 const ROOT_DIR = path.resolve(__dirname, '../../');
 
+/**
+ * @return {Promise<string[]>}
+ */
+async function getPathsToDelete() {
+  const [, below] = splitIgnoreListByHeader(
+    await fs.readFile(ignoreFile, 'utf8')
+  );
+  return (
+    below
+      .split('\n')
+      // Comments and empty lines
+      .filter((line) => line.trim().length > 0 && !line.startsWith('#'))
+      // Recursive globs
+      .map((line) => (line.startsWith('/') ? line.substr(1) : `**/${line}`))
+  );
+}
+
 /**
  * Cleans up various cache and output directories. Optionally cleans up inner
  * node_modules package directories, or excludes some directories from deletion.
  * @return {Promise<void>}
  */
 async function clean() {
-  const pathsToDelete = [
-    // Local cache directories
-    // Keep this list in sync with .gitignore, .eslintignore, and .prettierignore
-    '.babel-cache',
-    '.css-cache',
-    '.jss-cache',
-    '.pre-closure-cache',
-
-    // Output directories
-    // Keep this list in sync with .gitignore, .eslintignore, and .prettierignore
-    '.amp-dep-check',
-    'build',
-    'build-system/dist',
-    'build-system/server/new-server/transforms/dist',
-    'build-system/tasks/performance/cache',
-    'build-system/tasks/performance/results.json',
-    'dist',
-    'dist.3p',
-    'dist.tools',
-    'export',
-    'examples/storybook',
-    'extensions/**/dist',
-    'release',
-    'result-reports',
-    'test/coverage',
-    'test/coverage-e2e',
-    'validator/**/dist',
-    'validator/export',
+  const pathsToDeleteFromIgnore = [
+    ...(await getPathsToDelete()),
+    '!**/third_party',
+    '!**/node_modules',
   ];
+  const pathsToDelete = [];
   if (argv.include_subpackages) {
     pathsToDelete.push('**/node_modules', '!node_modules');
   }
-  // User configuration files
-  // Keep this list in sync with .gitignore, .eslintignore, and .prettierignore
-  const customConfigs = [
-    'build-system/global-configs/custom-config.json',
-    'build-system/global-configs/custom-flavors-config.json',
-  ];
-  if (argv.include_custom_configs) {
-    pathsToDelete.push(...customConfigs);
-  } else {
-    for (const customConfig of customConfigs) {
-      if (fs.existsSync(customConfig)) {
-        log(yellow('Skipping path:'), cyan(customConfig));
+
+  // Ignore user configuration files if flag is not set.
+  const ignoredCustomConfigPaths = pathsToDeleteFromIgnore.filter((path, i) => {
+    if (/build-system\/global-configs\/custom.*\.json/.test(path)) {
```

Comment:
This pattern just happens to work right now, but may break if someone were to add a new config file that doesn't match the pattern. Also, I'm not convinced that the regex logic is all that readable or maintainable.

Instead, can we restore the `customConfigs` list and check if `path` is included in that list? (Might help to include a comment that points to https://github.com/ampproject/amphtml/blob/main/build-system/global-configs/README.md.)